### PR TITLE
C# SDK and .NET Core Example

### PR DESCRIPTION
**Pull request by TimoSchmechel**
I added a full implementation of the Agones REST API in C#. I saw there was already an existing Unity SDK however it was using specific Unity webrequests which wont work in other C# codebases. 
This SDK works in Unity as well as .NET Core and Framework. I includes the visual studio files, if you want me to remove them and just keep it as code files only I am happy to remove them. 
I have included a basic .NET Core example which uses all the features of the API.

### CODE REVIEW

**Comment by FelixKlauke** [comment]
Nice work! Can you please squash your commits into a single one? You may also take a look at [build-sdk-images](https://github.com/googleforgames/agones/tree/master/build/build-sdk-images). Even if code generation isn't needed, the tests should be run.

AFAIK gRPC supports C#, is there any reason you cannot use it? It's the preferred way of implementation.

File: sdks/csharp/AgonesClient/AgonesClient.cs [hunk 1]
```diff
@@ -0,0 +1,225 @@
+﻿using Agones.Models;
```
Comment (FelixKlauke) [inline sdks/csharp/AgonesClient/AgonesClient.cs]: We need a license header here.

File: sdks/csharp/AgonesClient/ConsoleLogger.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
+﻿using System;
```
Comment (FelixKlauke) [inline sdks/csharp/AgonesClient/ConsoleLogger.cs]: We need a license header here.

File: sdks/csharp/AgonesClient/GameServerInfo.cs [hunk 3]
```diff
@@ -0,0 +1,107 @@
+﻿using Agones.Utility;
```
Comment (FelixKlauke) [inline sdks/csharp/AgonesClient/GameServerInfo.cs]: We need a license header here.

File: sdks/csharp/AgonesClient/ILogger.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
```
Comment (FelixKlauke) [inline sdks/csharp/AgonesClient/ILogger.cs]: We need a license header here.

File: sdks/csharp/AgonesTestCore/Program.cs [hunk 5]
```diff
@@ -0,0 +1,61 @@
+﻿using Agones;
```
Comment (FelixKlauke) [inline sdks/csharp/AgonesTestCore/Program.cs]: This is just a local manual test class. Consider replacing this with real (unit) tests.

File: sdks/csharp/AgonesClient/CustomUnixDateTimeConverter.cs [hunk 6]
```diff
@@ -0,0 +1,29 @@
+﻿using Newtonsoft.Json;
```
Comment (FelixKlauke) [inline sdks/csharp/AgonesClient/CustomUnixDateTimeConverter.cs]: We need a license header here.

**Comment by TimoSchmechel** [comment]
@FelixKlauke The main reason I didnt use gRPC generation is because I had never used it before and have decent experience writing a REST client manually. As I am actively using this client in the game Im developing I felt it was faster for me to go down that route. Happy to look at generation though. 
I'll look through the testing, license header and other issues you mentioned over the next few days.

**Comment by TimoSchmechel** [comment]
Just created an Issue regarding this here #884

**Comment by markmandel** [comment]
I would agree with @FelixKlauke - C# has pretty good gRPC support, and where possible, we should aim for a gRPC implementation. It will be easier to maintain, has more features, and have better performance.

We only go REST when we have to/its a better solution.

**Comment by markmandel** [comment]
@googlebot rescan

**Comment by markmandel** [comment]
@googlebot rescan

File: .gitignore [hunk 7]
```diff
@@ -1,34 +1,37 @@
-# Copyright 2017 Google LLC All Rights Reserved.
```
Comment (aLekSer) [inline .gitignore]: I think this file showed up just because you changed line endings from \n to \r\n, from Linux style to Win.

**Comment by aLekSer** [comment]
Also regarding VS proj files in the PR. Cmake supports C # project files generation. Would be useful if multiple IDE would be used with this SDK.
Makefile as in `build-sdk-images` would be useful to understand how to build the project if possible from CLI.

**Comment by TimoSchmechel** [comment]
Alright seems like you're pretty adamant about going for gRPC, which makes sense as its easier to maintain going forward. Happy to reimplement using that some time next week when I get time.

How well does it support streams? Specifically the watch gameserver case

**Comment by markmandel** [comment]
> How well does it support streams? Specifically the watch gameserver case

About a million times easier :smile: single/biderectional streaming is wher gRPC really shines (among other things)

**Comment by markmandel** [comment]
Should we close this PR, and reopen for a new gRPC based one?

**Comment by TimoSchmechel** [comment]
Yup makes the most sense

**Pull request closed without merging**